Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +370,124 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
-}
+    if (!has_Dirichlet) return;
+
+    // loop over all mesh refinement levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr   = phi[lev]->array(mfi);
+            // Extract tileboxes for which to loop
+            const Box& tb  = mfi.tilebox( phi[lev]->ixType().toIntVect() );
+
+            const auto lo = lbound(phi_arr);
+            const auto hi = ubound(phi_arr);
+
+            amrex::ParallelFor( tb,
```

Comment:
Is it possible to subset the tileboxes to just those that lie on a (relevant) boundary before dispatching a loop over all of them to the GPU? In a square-ish domain, most of the boxes will have no boundary, and so including them would just add overhead.

Reply:
Quickly returning unused threads is actually really cheap on GPU :-)

You can select the boundaries, but then we need to spawn more kernels at the moment (see e.g. https://github.com/ECP-WarpX/WarpX/pull/1766#discussion_r589700523). Spawning more kernels comes with a significant overhead compared to quickly returning unused threads.